ratelimit/rate_limiter.go

```diff
@@ -0,0 +1,186 @@
+}
+
+// Start starts two background processes required for the RateLimiter to function. One that
+// stores it's state to the DB at a checkpoint interval, and another that clears accrued
+// grants when the UTC day time window elapses.
+func (r *RateLimiter) Start(ctx context.Context, checkpointInterval time.Duration) error {
+	r.startMutex.Lock()
+	if r.wasStartedOnce {
+		r.startMutex.Unlock()
+		return errors.New("Can only start RateLimiter once per instance")
+	}
+	r.wasStartedOnce = true
+	r.startMutex.Unlock()
+
+	// Start 24hr UTC accrued grants resetter
+	go func() {
+		for {
+			now := r.aClock.Now()
+			currentUTCCheckpoint := getUTCMidnightOfDate(now)
+			nextUTCCheckpoint := currentUTCCheckpoint.Add(24 * time.Hour)
+			untilNextUTCCheckpoint := nextUTCCheckpoint.Sub(r.aClock.Now())
+			select {
+			case <-ctx.Done():
+				return
+			case <-r.aClock.After(untilNextUTCCheckpoint):
+				// Compute how many grants have accrued and gone unused and remove that
+				// many from the bucket so that it starts empty for the next 24hr period
+				r.mu.Lock()
+				accruedGrants := r.maxRequestsPer24Hrs - r.grantedInLast24hrsUTC
+				if err := r.twentyFourHourLimiter.WaitN(ctx, accruedGrants); err != nil {
+					// Since we never set n to exceed the burst size, an error will only
+					// occur if the context is cancelled or it's deadline is exceeded. In
+					// these cases, we simply return so that this go-routine exits.
+					// From docs: "It returns an error if n exceeds the Limiter's burst
+					// size, the Context is canceled, or the expected wait time exceeds the
+					// Context's Deadline."
+					// Source: https://godoc.org/golang.org/x/time/rate#Limiter.WaitN
+					r.mu.Unlock()
+					return
+				}
+				r.currentUTCCheckpoint = nextUTCCheckpoint
+				r.grantedInLast24hrsUTC = 0
+				r.mu.Unlock()
+			}
+		}
+	}()
+
+	ticker := time.NewTicker(checkpointInterval)
+	for {
+		select {
+		case <-ctx.Done():
+			ticker.Stop()
+			return nil
+		case <-ticker.C:
+			// Store grants issued and current UTC checkpoint to DB
+			r.mu.Lock()
+			err := r.meshDB.UpdateMetadata(func(metadata meshdb.Metadata) meshdb.Metadata {
+				metadata.StartOfCurrentUTCDay = r.currentUTCCheckpoint
+				metadata.EthRPCRequestsSentInCurrentUTCDay = r.grantedInLast24hrsUTC
+				return metadata
+			})
+			r.mu.Unlock()
+			if err != nil {
+				if err == leveldb.ErrClosed {
+					// We can't continue if the database is closed. Stop the rateLimiter and
+					// return an error.
+					ticker.Stop()
+					return err
+				}
+				log.WithError(err).Error("rateLimiter.Start() error encountered while updating metadata in DB")
+			}
+		}
+	}
+}
+
+func getUTCMidnightOfDate(date time.Time) time.Time {
+	return time.Date(date.Year(), date.Month(), date.Day(), 0, 0, 0, 0, time.UTC)
+}
+
+// Wait blocks until the rateLimiter allows for another request to be sent
+func (r *RateLimiter) Wait(ctx context.Context) error {
+	if err := r.twentyFourHourLimiter.Wait(ctx); err != nil {
```

Comment:
I'm not sure if this behaves the way that we want it to. What happens if `r.twentyFourHourLimiter.Wait(ctx)` returns immediately but we have to wait for `r.perSecondLimiter.Wait(ctx)`? If it's close to midnight UTC it seems like this could result in more than 100k requests sent over a 24 hour period.

Reply:
Fair. We've decided offline to add in a buffer so that we don't need to worry about this edge-case.